src/workerd/api/sql-test.js

```diff
@@ -0,0 +1,81 @@
+function assertEqual(a, b) {
+  if (a !== b) {
+    throw new Error(a + " !== " + b);
+  }
+}
+
+function requireException(callback, expectStr) {
+  try {
+    callback();
+  } catch (err) {
+    const errStr = `${err}`;
+    if (!errStr.includes(expectStr)) {
+      throw new Error(`Got unexpected exception '${errStr}', expected: '${expectStr}'`);
+    }
+    return;
+  }
+  throw new Error(`Expected exception '${expectStr}' but none was thrown`);
+}
+
+export default {
+  async test(ctrl, env, ctx) {
+
+    const sql = new SqlDatabase();
+
+    // Test numeric results
+    const resultNumber = sql.exec("SELECT 123", null).results;
+    assertEqual(resultNumber.length, 1);
+    assertEqual(resultNumber[0]["123"], 123);
+
+    // Test string results
+    const resultStr = sql.exec("SELECT 'hello'", null).results;
```

Comment:
If we're going to bake this in as a runtime API, we should absolutely make sure we also bake in something like https://www.npmjs.com/package/@nearform/sql to provide some protection against SQL injection.

Reply:
Unfortunately I don't believe so. I'm also not convinced we'd want that as it would likely be a surprising dx limitation.

Reply:
I guess it would also be a problem for D1, which receives queries from the network. But we could offer a separate `execUnsafe()` or something that skips the check.

If we can't have V8 tell us if the string is computed, another possibility would be to make people wrap strings in some sort of "prepared statement" objects at startup time -- where creating the objects after startup is disallowed. This might even give us some interesting optimization opportunities some day where we could pre-parse the strings (though there's no simple way to do that now since sqlite only lets you parse SQL in the context of a specific database). That said, the DX of this may be not so great.

I wish we could introduce our own syntax like JS regex syntax that can be written inline but would still be parsed at startup...

Reply:
> ... another possibility would be to make people wrap strings in some sort of "prepared statement" object...

We can, actually.

JS tagged template literals do not have to return strings and they are evaluated lexically. So, for instance, we could define something equivalent to...

```
function SQL(strings, ...args) {
  // process the inputs here...
  return new PreparedStatement(...);
}

const statement = SQL`SELECT ...`;
```

Because the template literal tag returns an instance of `PreparedStatement`, `statement` will be that object instead of a string. We could easily make it so that user code itself has no way of calling "new PreparedStatement" and we can properly process the inputs to the template literal tag function to ensure proper escaping.

The DX of this would be familiar to many JS developers.